pkg/grpc/jwt_authenticator_test.go

```diff
@@ -0,0 +1,131 @@
+package grpc_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/buildbarn/bb-storage/internal/mock"
+	bb_grpc "github.com/buildbarn/bb-storage/pkg/grpc"
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/require"
+	"gopkg.in/square/go-jose.v2"
+	"gopkg.in/square/go-jose.v2/jwt"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+func TestJWTAuthenticator(t *testing.T) {
+	ctrl, ctx := gomock.WithContext(context.Background(), t)
+	defer ctrl.Finish()
+	clock := mock.NewMockClock(ctrl)
+
+	symmetricKey := []byte("0123456789ABCDEF")
+
+	jwtKey := bb_grpc.JWTKeyConfig{
+		Key: symmetricKey,
+	}
+
+	signer := mustMakeSigner(jose.HS256, symmetricKey)
+
+	authenticator := bb_grpc.NewJWTAuthenticator(jwtKey, clock)
+
+	t.Run("NoGRPC", func(t *testing.T) {
+		// Authenticator is used outside of gRPC, meaning it cannot
+		// extract request metadata.
+		require.Equal(
+			t,
+			status.Error(codes.Unauthenticated, "Connection was not established using gRPC"),
+			authenticator.Authenticate(ctx))
+	})
+
+	t.Run("NoAuthorizationMetadata", func(t *testing.T) {
+		// Should deny authentication if no `authorization` header is present.
+		md := metadata.MD{}
+		require.Equal(
+			t,
+			status.Error(codes.Unauthenticated, "Missing authorization header"),
+			authenticator.Authenticate(metadata.NewIncomingContext(ctx, md)),
+		)
+	})
+
+	t.Run("HasAuthorizationMetadataKeyButNoValues", func(t *testing.T) {
+		// Should deny authentication if `authorization` header is present but has no values.
+		md := metadata.MD{
+			"authorization": nil,
+		}
+		require.Equal(
+			t,
+			status.Error(codes.Unauthenticated, "Missing authorization header"),
+			authenticator.Authenticate(metadata.NewIncomingContext(ctx, md)),
+		)
+	})
+
+	t.Run("HasAuthorizationMetadataKeyButMultipleValues", func(t *testing.T) {
+		// Should deny authentication if `authorization` header is present and has multiple values.
+		md := metadata.Pairs("authorization", "hello", "authorization", "world")
+		require.Equal(
+			t,
+			status.Error(codes.Unauthenticated, "Multiple authorization headers are not supported"),
```

Comment:
I still don't think that this is a wise strategy. What if the client provides both an `Authorization: Basic <...>` and `Authorization: Bearer <...>` header?

Reply:
I believe it is a judgment call. If you want to support the use case, then I'm fine with supporting it, although frankly I still don't see a case where a client would want to send two different types of authentication.

The code will have to be similar to what the AnyAuthenticator does: scan all of the auth headers and returns success if there any valid JWT token versus failing at the first auth header that does not authenticate as a JWT (which could be a non-JWT auth header like basic auth). The docs will need to clarify that behavior.